pkg/sources/redisstream/redisstream.go

```diff
@@ -0,0 +1,273 @@
+	redisclient "github.com/numaproj/numaflow/pkg/shared/clients/redis"
+	"github.com/numaproj/numaflow/pkg/shared/logging"
+	sharedutil "github.com/numaproj/numaflow/pkg/shared/util"
+	"github.com/numaproj/numaflow/pkg/watermark/fetch"
+	"github.com/numaproj/numaflow/pkg/watermark/processor"
+	"github.com/numaproj/numaflow/pkg/watermark/publish"
+	"github.com/numaproj/numaflow/pkg/watermark/store"
+	"github.com/numaproj/numaflow/pkg/watermark/wmb"
+	"go.uber.org/zap"
+)
+
+type redisStreamsSource struct {
+	// shared code to read from Redis Streams
+	*redisclient.RedisStreamsReader
+	// name of the pipeline
+	pipelineName string
+	// forwarder that writes the consumed data to destination
+	forwarder *forward.InterStepDataForward
+	// context cancel function
+	cancelfn context.CancelFunc
+	// source watermark publisher
+	sourcePublishWM publish.Publisher
+}
+
+type Option func(*redisStreamsSource) error
+
+// WithLogger is used to return logger information
+func WithLogger(l *zap.SugaredLogger) Option {
+	return func(o *redisStreamsSource) error {
+		o.Log = l
+		return nil
+	}
+}
+
+// WithReadTimeOut sets the read timeout
+func WithReadTimeOut(t time.Duration) Option {
+	return func(o *redisStreamsSource) error {
+		o.Options.ReadTimeOut = t
+		return nil
+	}
+}
+
+func New(
+	vertexInstance *dfv1.VertexInstance,
+	writers []isb.BufferWriter,
+	fsd forward.ToWhichStepDecider,
+	mapApplier applier.MapApplier,
+	fetchWM fetch.Fetcher,
+	publishWM map[string]publish.Publisher,
+	publishWMStores store.WatermarkStorer,
+	opts ...Option) (*redisStreamsSource, error) {
+
+	// create RedisClient to connect to Redis
+	vertexSpec := vertexInstance.Vertex.Spec
+	redisSpec := vertexSpec.Source.RedisStreams
+	redisClient, err := newRedisClient(redisSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	readerOpts := &redisclient.Options{
+		InfoRefreshInterval: time.Second,
+		ReadTimeOut:         time.Second,
+		CheckBackLog:        true,
+	}
+
+	// RedisStreamsReader handles reading from Redis Streams
+	redisStreamsReader := &redisclient.RedisStreamsReader{
+		Name:        vertexSpec.Name,
+		Stream:      redisSpec.Stream,
+		Group:       redisSpec.ConsumerGroup,
+		Consumer:    fmt.Sprintf("%s-%v", vertexSpec.Name, vertexInstance.Replica),
+		RedisClient: redisClient,
+		Options:     *readerOpts,
+		Metrics: redisclient.Metrics{
+			ReadErrorsInc: func() {
+				redisStreamsSourceReadErrors.With(map[string]string{metrics.LabelVertex: vertexSpec.Name, metrics.LabelPipeline: vertexSpec.PipelineName}).Inc()
+			},
+			ReadsAdd: func(count int) {
+				redisStreamsSourceReadCount.With(map[string]string{metrics.LabelVertex: vertexSpec.Name, metrics.LabelPipeline: vertexSpec.PipelineName}).Add(float64(count))
+			},
+			AcksAdd: func(count int) {
+				redisStreamsSourceAckCount.With(map[string]string{metrics.LabelVertex: vertexSpec.Name, metrics.LabelPipeline: vertexSpec.PipelineName}).Add(float64(count))
+			},
+		},
+	}
+	redisStreamsReader.Log = logging.NewLogger()
```

Comment:
Can use the log from the input `opts`, right?

Reply:
yes, I can instantiate this if not passed in as an option